File: yujung/프로그래머스/기능개발.cpp
#include <string>
#include <vector>
#include<cmath>
#include<queue>
using namespace std;

queue<int> q;
vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    int cnt=1;
    int max=0;
    for(int i=0; i<progresses.size();i++){
    progresses[i]=100-progresses[i];
    progresses[i]=ceil((float)progresses[i]/speeds[i]);
        q.push(progresses[i]);
    }
    while(!q.empty()){
        max=q.front();
        q.pop();
        while(!q.empty()&&max>=q.front()){
            q.pop();
            cnt++;
        }
        answer.push_back(cnt);
        cnt=1;
    }
    
    return answer;
}

File: seoyeon/프로그래머스/알고리즘 고득점 Kit/스택/큐/기능개발.py
#1. 90.0/100
#progress:작업진도 ,speeds:개발속도
# def solution(progresses, speeds):
#     answer = []
    
#     end_time = [0 for _ in range(len(progresses))]
    
#     for i in range(len(progresses)):
#         if (100-progresses[i])//speeds[i]==0:
#             end_time[i]=(100-progresses[i])//speeds[i]
#         else:
#             end_time[i]=(100-progresses[i])//speeds[i] + 1
            
#     visited=[False for _ in range(len(progresses))]
#     for i in range(len(progresses)):
#         if visited[i]==True:
#             continue
#         start = i
#         ans = 1
#         for j in range(i+1,len(progresses)):
#             if end_time[j]<=end_time[start]:
#                 ans += 1
#                 visited[j]=True
#             else:
#                 break
#         answer.append(ans)
        
#     return answer

#2. 100/100
#progress:작업진도 ,speeds:개발속도
def solution(progresses, speeds):
    answer = []
    
    end_time = [0 for _ in range(len(progresses))]
    
    for i in range(len(progresses)):
        if (100-progresses[i])%speeds[i]==0: #이 부분 수정
            end_time[i]=(100-progresses[i])//speeds[i]
        else:
            end_time[i]=(100-progresses[i])//speeds[i] + 1
    
    visited=[False for _ in range(len(progresses))]
    for i in range(len(progresses)):
        if visited[i]==True:
            continue

        start = i
        ans = 1
        for j in range(i+1,len(progresses)):
            if end_time[j]<=end_time[start]:
                ans += 1
                visited[j]=True
            else:
                break
        answer.append(ans)
        
    return answer